// @uipath/uipath-typescript/dist/document-understanding/index.mjs
// Auto-generated from the OpenAPI spec — do not edit manually.
var DocumentActionPriority;
(function (DocumentActionPriority) {
    DocumentActionPriority["Low"] = "Low";
    DocumentActionPriority["Medium"] = "Medium";
    DocumentActionPriority["High"] = "High";
    DocumentActionPriority["Critical"] = "Critical";
})(DocumentActionPriority || (DocumentActionPriority = {}));
var DocumentActionStatus;
(function (DocumentActionStatus) {
    DocumentActionStatus["Unassigned"] = "Unassigned";
    DocumentActionStatus["Pending"] = "Pending";
    DocumentActionStatus["Completed"] = "Completed";
})(DocumentActionStatus || (DocumentActionStatus = {}));
var DocumentActionType;
(function (DocumentActionType) {
    DocumentActionType["Validation"] = "Validation";
    DocumentActionType["Classification"] = "Classification";
})(DocumentActionType || (DocumentActionType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var ProjectProperties;
(function (ProjectProperties) {
    ProjectProperties["IsPredefined"] = "IsPredefined";
    ProjectProperties["SupportsTags"] = "SupportsTags";
    ProjectProperties["SupportsVersions"] = "SupportsVersions";
    ProjectProperties["IsSplittingEnabled"] = "IsSplittingEnabled";
})(ProjectProperties || (ProjectProperties = {}));
var ProjectType;
(function (ProjectType) {
    ProjectType["Classic"] = "Classic";
    ProjectType["Modern"] = "Modern";
    ProjectType["IXP"] = "IXP";
    ProjectType["Unknown"] = "Unknown";
})(ProjectType || (ProjectType = {}));
var ResourceStatus;
(function (ResourceStatus) {
    ResourceStatus["Unknown"] = "Unknown";
    ResourceStatus["NotStarted"] = "NotStarted";
    ResourceStatus["ExportInProgress"] = "ExportInProgress";
    ResourceStatus["ExportCompleted"] = "ExportCompleted";
    ResourceStatus["InProgress"] = "InProgress";
    ResourceStatus["Suspended"] = "Suspended";
    ResourceStatus["Resuming"] = "Resuming";
    ResourceStatus["Available"] = "Available";
    ResourceStatus["Error"] = "Error";
    ResourceStatus["Deleting"] = "Deleting";
})(ResourceStatus || (ResourceStatus = {}));
var ResourceType;
(function (ResourceType) {
    ResourceType["Specialized"] = "Specialized";
    ResourceType["Generative"] = "Generative";
    ResourceType["Custom"] = "Custom";
    ResourceType["Unknown"] = "Unknown";
})(ResourceType || (ResourceType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var MarkupType;
(function (MarkupType) {
    MarkupType["Unknown"] = "Unknown";
    MarkupType["Circled"] = "Circled";
    MarkupType["Underlined"] = "Underlined";
    MarkupType["Strikethrough"] = "Strikethrough";
})(MarkupType || (MarkupType = {}));
var ProcessingSource;
(function (ProcessingSource) {
    ProcessingSource["Unknown"] = "Unknown";
    ProcessingSource["Ocr"] = "Ocr";
    ProcessingSource["Pdf"] = "Pdf";
    ProcessingSource["PlainText"] = "PlainText";
    ProcessingSource["PdfAndOcr"] = "PdfAndOcr";
})(ProcessingSource || (ProcessingSource = {}));
var Rotation;
(function (Rotation) {
    Rotation["None"] = "None";
    Rotation["Rotated90"] = "Rotated90";
    Rotation["Rotated180"] = "Rotated180";
    Rotation["Rotated270"] = "Rotated270";
    Rotation["Other"] = "Other";
})(Rotation || (Rotation = {}));
var SectionType;
(function (SectionType) {
    SectionType["Vertical"] = "Vertical";
    SectionType["Paragraph"] = "Paragraph";
    SectionType["Header"] = "Header";
    SectionType["Footer"] = "Footer";
    SectionType["Table"] = "Table";
})(SectionType || (SectionType = {}));
var TextType;
(function (TextType) {
    TextType["Unknown"] = "Unknown";
    TextType["Text"] = "Text";
    TextType["Checkbox"] = "Checkbox";
    TextType["Handwriting"] = "Handwriting";
    TextType["Barcode"] = "Barcode";
    TextType["QRcode"] = "QRcode";
    TextType["Stamp"] = "Stamp";
    TextType["Logo"] = "Logo";
    TextType["Circle"] = "Circle";
    TextType["Underline"] = "Underline";
    TextType["Cut"] = "Cut";
})(TextType || (TextType = {}));
var WordGroupType;
(function (WordGroupType) {
    WordGroupType["Sentence"] = "Sentence";
    WordGroupType["TableCell"] = "TableCell";
    WordGroupType["TableRowEnd"] = "TableRowEnd";
    WordGroupType["Heading"] = "Heading";
    WordGroupType["Other"] = "Other";
})(WordGroupType || (WordGroupType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var ModelKind;
(function (ModelKind) {
    ModelKind["Classifier"] = "Classifier";
    ModelKind["Extractor"] = "Extractor";
})(ModelKind || (ModelKind = {}));
var ModelType;
(function (ModelType) {
    ModelType["IXP"] = "IXP";
    ModelType["Modern"] = "Modern";
    ModelType["Predefined"] = "Predefined";
})(ModelType || (ModelType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var ClassifierDocumentTypeType;
(function (ClassifierDocumentTypeType) {
    ClassifierDocumentTypeType["FormsAi"] = "FormsAi";
    ClassifierDocumentTypeType["SemiStructuredAi"] = "SemiStructuredAi";
    ClassifierDocumentTypeType["Helix"] = "Helix";
    ClassifierDocumentTypeType["Unknown"] = "Unknown";
})(ClassifierDocumentTypeType || (ClassifierDocumentTypeType = {}));
var CreateTaskPriority;
(function (CreateTaskPriority) {
    CreateTaskPriority["Low"] = "Low";
    CreateTaskPriority["Medium"] = "Medium";
    CreateTaskPriority["High"] = "High";
    CreateTaskPriority["Critical"] = "Critical";
})(CreateTaskPriority || (CreateTaskPriority = {}));
var GptFieldType;
(function (GptFieldType) {
    GptFieldType["Address"] = "Address";
    GptFieldType["Boolean"] = "Boolean";
    GptFieldType["Date"] = "Date";
    GptFieldType["Name"] = "Name";
    GptFieldType["Number"] = "Number";
    GptFieldType["Text"] = "Text";
})(GptFieldType || (GptFieldType = {}));
var ValidationDisplayMode;
(function (ValidationDisplayMode) {
    ValidationDisplayMode["Classic"] = "Classic";
    ValidationDisplayMode["Compact"] = "Compact";
})(ValidationDisplayMode || (ValidationDisplayMode = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var ResultsDataSource;
(function (ResultsDataSource) {
    ResultsDataSource["Automatic"] = "Automatic";
    ResultsDataSource["Manual"] = "Manual";
    ResultsDataSource["ManuallyChanged"] = "ManuallyChanged";
    ResultsDataSource["Defaulted"] = "Defaulted";
    ResultsDataSource["External"] = "External";
})(ResultsDataSource || (ResultsDataSource = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var ComparisonOperator;
(function (ComparisonOperator) {
    ComparisonOperator["Equals"] = "Equals";
    ComparisonOperator["NotEquals"] = "NotEquals";
    ComparisonOperator["Greater"] = "Greater";
    ComparisonOperator["Less"] = "Less";
    ComparisonOperator["GreaterOrEqual"] = "GreaterOrEqual";
    ComparisonOperator["LessOrEqual"] = "LessOrEqual";
})(ComparisonOperator || (ComparisonOperator = {}));
var Criticality;
(function (Criticality) {
    Criticality["Must"] = "Must";
    Criticality["Should"] = "Should";
})(Criticality || (Criticality = {}));
var FieldType;
(function (FieldType) {
    FieldType["Text"] = "Text";
    FieldType["Number"] = "Number";
    FieldType["Date"] = "Date";
    FieldType["Name"] = "Name";
    FieldType["Address"] = "Address";
    FieldType["Keyword"] = "Keyword";
    FieldType["Set"] = "Set";
    FieldType["Boolean"] = "Boolean";
    FieldType["Table"] = "Table";
    FieldType["Internal"] = "Internal";
    FieldType["FieldGroup"] = "FieldGroup";
    FieldType["MonetaryQuantity"] = "MonetaryQuantity";
})(FieldType || (FieldType = {}));
var LogicalOperator;
(function (LogicalOperator) {
    LogicalOperator["AND"] = "AND";
    LogicalOperator["OR"] = "OR";
})(LogicalOperator || (LogicalOperator = {}));
var RuleType;
(function (RuleType) {
    RuleType["Mandatory"] = "Mandatory";
    RuleType["PossibleValues"] = "PossibleValues";
    RuleType["Regex"] = "Regex";
    RuleType["StartsWith"] = "StartsWith";
    RuleType["EndsWith"] = "EndsWith";
    RuleType["FixedLength"] = "FixedLength";
    RuleType["IsNumeric"] = "IsNumeric";
    RuleType["IsDate"] = "IsDate";
    RuleType["IsEmail"] = "IsEmail";
    RuleType["Contains"] = "Contains";
    RuleType["Expression"] = "Expression";
    RuleType["TableExpression"] = "TableExpression";
    RuleType["IsEmpty"] = "IsEmpty";
})(RuleType || (RuleType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
var ActionStatus;
(function (ActionStatus) {
    ActionStatus["Unassigned"] = "Unassigned";
    ActionStatus["Pending"] = "Pending";
    ActionStatus["Completed"] = "Completed";
})(ActionStatus || (ActionStatus = {}));

var index = /*#__PURE__*/Object.freeze({
    __proto__: null,
    get ActionStatus () { return ActionStatus; },
    get ClassifierDocumentTypeType () { return ClassifierDocumentTypeType; },
    get ComparisonOperator () { return ComparisonOperator; },
    get CreateTaskPriority () { return CreateTaskPriority; },
    get Criticality () { return Criticality; },
    get DocumentActionPriority () { return DocumentActionPriority; },
    get DocumentActionStatus () { return DocumentActionStatus; },
    get DocumentActionType () { return DocumentActionType; },
    get FieldType () { return FieldType; },
    get GptFieldType () { return GptFieldType; },
    get LogicalOperator () { return LogicalOperator; },
    get MarkupType () { return MarkupType; },
    get ModelKind () { return ModelKind; },
    get ModelType () { return ModelType; },
    get ProcessingSource () { return ProcessingSource; },
    get ProjectProperties () { return ProjectProperties; },
    get ProjectType () { return ProjectType; },
    get ResourceStatus () { return ResourceStatus; },
    get ResourceType () { return ResourceType; },
    get ResultsDataSource () { return ResultsDataSource; },
    get Rotation () { return Rotation; },
    get RuleType () { return RuleType; },
    get SectionType () { return SectionType; },
    get TextType () { return TextType; },
    get ValidationDisplayMode () { return ValidationDisplayMode; },
    get WordGroupType () { return WordGroupType; }
});

export { index as DuFramework };
